PR: https://github.com/odoo/odoo/pull/212679

From: 2713876dbc70d3984e584a9037a2206dcda4e84a
From: Arnold Moyaux
Date: 2025-09-07 15:50:27

Breaking data model changes scores: del:4 + add:5, change matches:
+    production_ids = fields.One2many('mrp.production', 'production_group_id', string='Productions')
+    child_ids = fields.Many2many(
+    parent_ids = fields.Many2many(
-    procurement_group_id = fields.Many2one(
+    reference_ids = fields.Many2many(
+    production_ids = fields.One2many(
-    _inherit = 'procurement.group'
-    mrp_production_ids = fields.One2many('mrp.production', 'procurement_group_id')

Total Changes: 373

[REF] stock, *: add reference

The model `procurement.group` is replaced by a more flexible one:
`stock.reference`. The first one was somehow not well understood and a
bit limited by its many2one links to other business model (Sales Order,
Purchase Order, ...). The reference is made of a name and many2many to
all the documents integrated with Inventory. It's purpose has been
shrank to the minimal "make links between object". For instance
opening the manufacturing order created for a MTO sale. The merge of
stock move is delegated to existing parameters (`sale_id`,
created_purchase_line_ids`, ...)

The `mrp.production` needed an extra effort to group them according to
inheritance relationships. This commit add a new object to group
productions together either in a backorder relation or in a child-parent
one. This allow for instance to have a Sale Order linked to the MTO
product of the sold product and the production them selves to be linked
to the SO *and* the potential child MOs.

The procurement group propagation, before set on the Stock Rule, allowing
the stock moves to be merged in an existing picking or not has been
removed and replaced (for the creation of purchase order only) by a new
parameter on `res.partner` allowing the group of Request for Quotation
base on the scheduled date of the potential merge candidate (never, same
week, same month, always). All other merge scenarios should use batches
and waves transfers.

closes odoo/odoo#212679

Task: 4279803
Related: odoo/enterprise#87170
Related: odoo/upgrade#8326
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp/models/__init__.py
+++ b/addons/mrp/models/__init__.py
@@ -17,6 +17,7 @@ from . import stock_move_line
 from . import stock_orderpoint
 from . import stock_picking
 from . import stock_lot
+from . import stock_reference
 from . import stock_rule
 from . import stock_scrap
 from . import stock_warehouse

--- a/addons/mrp/models/mrp_production.py
+++ b/addons/mrp/models/mrp_production.py
@@ -21,6 +21,20 @@ from odoo.addons.stock.models.stock_move import PROCUREMENT_PRIORITIES
 SIZE_BACK_ORDER_NUMERING = 3
 
 
+class MrpProductionGroup(models.Model):
+    _name = 'mrp.production.group'
+    _description = 'Production Group'
+
+    name = fields.Char('Name', required=True, index='btree')
+    production_ids = fields.One2many('mrp.production', 'production_group_id', string='Productions')
+    child_ids = fields.Many2many(
+        'mrp.production.group', 'mrp_production_group_rel', 'parent_group_id', 'child_group_id',
+        string='Child Manufacturing Orders')
+    parent_ids = fields.Many2many(
+        'mrp.production.group', 'mrp_production_group_rel', 'child_group_id', 'parent_group_id',
+        string='Parent Manufacturing Orders')
+
+
 class MrpProduction(models.Model):
     """ Manufacturing Orders """
     _name = 'mrp.production'
@@ -63,6 +77,8 @@ class MrpProduction(models.Model):
         domain="[('type', '=', 'consu')]",
         compute='_compute_product_id', store=True, copy=True, precompute=True,
         readonly=False, required=True, check_company=True)
+    production_group_id = fields.Many2one('mrp.production.group', 'Production Group', index=True, copy=False)
+
     product_variant_attributes = fields.Many2many('product.template.attribute.value', related='product_id.product_template_attribute_value_ids')
     valid_product_template_attribute_line_ids = fields.Many2many(related='product_tmpl_id.valid_product_template_attribute_line_ids')
     never_product_template_attribute_value_ids = fields.Many2many(
@@ -207,9 +223,9 @@ class MrpProduction(models.Model):
         index=True, required=True)
 
     qty_produced = fields.Float(compute="_get_produced_qty", string="Quantity Produced")
-    procurement_group_id = fields.Many2one(
-        'procurement.group', 'Procurement Group',
-        copy=False, index='btree_not_null')
+    reference_ids = fields.Many2many(
+        'stock.reference', 'stock_reference_production_rel', 'production_id', 'reference_id', 'References',
+    )
     product_description_variants = fields.Char('Custom Description')
     orderpoint_id = fields.Many2one('stock.warehouse.orderpoint', 'Orderpoint', copy=False, index='btree_not_null')
     propagate_cancel = fields.Boolean(
@@ -282,22 +298,20 @@ class MrpProduction(models.Model):
         'The quantity to produce must be positive!',
     )
 
-    @api.depends('procurement_group_id.stock_move_ids.created_production_id.procurement_group_id.mrp_production_ids',
-                 'procurement_group_id.stock_move_ids.move_orig_ids.created_production_id.procurement_group_id.mrp_production_ids')
+    @api.depends('production_group_id.child_ids.production_ids')
     def _compute_mrp_production_child_count(self):
         for production in self:
             production.mrp_production_child_count = len(production._get_children())
 
-    @api.depends('procurement_group_id.mrp_production_ids.move_dest_ids.group_id.mrp_production_ids',
-                 'procurement_group_id.stock_move_ids.move_dest_ids.group_id.mrp_production_ids')
+    @api.depends('production_group_id.parent_ids.production_ids')
     def _compute_mrp_production_source_count(self):
         for production in self:
             production.mrp_production_source_count = len(production._get_sources())
 
-    @api.depends('procurement_group_id.mrp_production_ids')
+    @api.depends('production_group_id.production_ids')
     def _compute_mrp_production_backorder(self):
         for production in self:
-            production.mrp_production_backorder_count = len(production.procurement_group_id.mrp_production_ids)
+            production.mrp_production_backorder_count = len(production.production_group_id.production_ids)
 
     @api.depends('company_id', 'bom_id')
     def _compute_picking_type_id(self):
@@ -480,19 +494,18 @@ class MrpProduction(models.Model):
                 ]
             })
 
-    @api.depends('procurement_group_id', 'procurement_group_id.stock_move_ids.group_id')
+    @api.depends('state')  # To get SAM moves after validation
     def _compute_picking_ids(self):
-        grouped_stock_pickings = self.env['stock.picking']._read_group(
-            domain=[('group_id', 'in', self.procurement_group_id.ids), ('group_id', '!=', False)],
-            aggregates=['id:recordset'],
-            groupby=['group_id'],
+        move_per_production_group = self.env['stock.move']._read_group(
+            [('production_group_id', 'in', self.production_group_id.ids)],
+            ['production_group_id'],
+            ['picking_id:recordset'],
         )
-        pickings_per_procurement_group = {
-            group_id.id: picking_ids.sorted() for group_id, picking_ids in grouped_stock_pickings
-        }
+        move_per_production = {}
+        for group, pickings in move_per_production_group:
+            move_per_production[group] = pickings
         for order in self:
-            order.picking_ids = pickings_per_procurement_group.get(order.procurement_group_id.id, [])
-            order.picking_ids |= order.move_raw_ids.move_orig_ids.picking_id
+            order.picking_ids = move_per_production.get(order.production_group_id, False)
             order.delivery_count = len(order.picking_ids)
 
     @api.depends('product_uom_id', 'product_qty', 'product_id.uom_id')
@@ -775,7 +788,10 @@ class MrpProduction(models.Model):
     @api.depends('product_id', 'bom_id', 'product_qty', 'product_uom_id', 'location_dest_id', 'date_finished', 'move_dest_ids', 'never_product_template_attribute_value_ids')
     def _compute_move_finished_ids(self):
         production_with_move_finished_ids_to_unlink_ids = OrderedSet()
+        ignored_mo_ids = self.env.context.get('ignore_mo_ids', [])
         for production in self:
+            if production.id in ignored_mo_ids:
+                continue
             if production.state != 'draft':
                 updated_values = {}
                 if production.date_finished:
@@ -997,12 +1013,19 @@ class MrpProduction(models.Model):
                     picking_type_id = self._get_default_picking_type_id(vals.get('company_id', self.env.company.id))
                     vals['picking_type_id'] = picking_type_id
                 vals['name'] = self.env['stock.picking.type'].browse(picking_type_id).sequence_id.next_by_id()
-            if not vals.get('procurement_group_id'):
-                procurement_group_vals = self._prepare_procurement_group_vals(vals)
-                vals['procurement_group_id'] = self.env["procurement.group"].create(procurement_group_vals).id
+            if not vals.get('production_group_id'):
+                vals['production_group_id'] = self.env["mrp.production.group"].create({'name': vals['name']}).id
         res = super().create(vals_list)
         # Make sure that the date passed in vals_list are taken into account and not modified by a compute
+        reference_vals_list = []
         for rec, vals in zip(res, vals_list):
+            (rec.move_raw_ids | rec.move_finished_ids).production_group_id = rec.production_group_id
+            if not rec.reference_ids:
+                reference_vals_list.append({
+                    'name': rec.name,
+                    'production_ids': [Command.set(rec.ids)],
+                    'move_ids': [Command.set(rec.move_raw_ids.ids + rec.move_finished_ids.ids)],
+                })
             if (rec.move_raw_ids
                 and rec.move_raw_ids[0].date
                 and vals.get('date_start')
@@ -1022,6 +1045,8 @@ class MrpProduction(models.Model):
                   and not vals.get('date_finished')):
                 # if no value is specified, do take the workorder duration (etc) into account
                 rec.move_finished_ids.write({'date': rec.date_finished})
+        if reference_vals_list:
+            self.env['stock.reference'].create(reference_vals_list)
         return res
 
     def unlink(self):
@@ -1160,10 +1185,6 @@ class MrpProduction(models.Model):
         ], limit=1).id
 
     def _get_move_finished_values(self, product_id, product_uom_qty, product_uom, operation_id=False, byproduct_id=False, cost_share=0):
-        group_orders = self.procurement_group_id.mrp_production_ids
-        move_dest_ids = self.move_dest_ids
-        if len(group_orders) > 1:
-            move_dest_ids |= group_orders[0].move_finished_ids.filtered(lambda m: m.product_id == self.product_id).move_dest_ids
         return {
             'product_id': product_id,
             'product_uom_qty': product_uom_qty,
@@ -1179,10 +1200,11 @@ class MrpProduction(models.Model):
             'production_id': self.id,
             'warehouse_id': self.location_dest_id.warehouse_id.id,
             'origin': self.product_id.partner_ref,
-            'group_id': self.procurement_group_id.id,
+            'reference_ids': self.reference_ids.ids,
             'propagate_cancel': self.propagate_cancel,
             'move_dest_ids': [(4, x.id) for x in self.move_dest_ids if not byproduct_id],
             'cost_share': cost_share,
+            'production_group_id': self.production_group_id.id,
         }
 
     def _get_moves_finished_values(self):
@@ -1249,6 +1271,7 @@ class MrpProduction(models.Model):
             - Manually added components, i.e. "default_" values in view
             - Moves from a copied MO, i.e. move.create
             - Existing moves during backorder creation """
+
         source_location = self.location_src_id
         data = {
             'sequence': bom_line.sequence if bom_line else 10,
@@ -1262,13 +1285,14 @@ class MrpProduction(models.Model):
             'location_id': source_location.id,
             'location_dest_id': self.product_id.with_company(self.company_id).property_stock_production.id,
             'raw_material_production_id': self.id,
+            'production_group_id': self.production_group_id.id,
             'company_id': self.company_id.id,
             'operation_id': operation_id,
             'procure_method': 'make_to_stock',
             'origin': self._get_origin(),
             'state': 'draft',
             'warehouse_id': source_location.warehouse_id.id,
-            'group_id': self.procurement_group_id.id,
+            'reference_ids': self.reference_ids.ids,
             'propagate_cancel': self.propagate_cancel,
             'manual_consumption': self.env['stock.move']._determine_is_manual_consumption(bom_line),
         }
@@ -1277,8 +1301,7 @@ class MrpProduction(models.Model):
     def _get_origin(self):
         origin = self.name
         if self.orderpoint_id and self.origin:
-            origin = self.origin.replace(
-                '%s - ' % (self.orderpoint_id.display_name), '')
+            origin = self.origin
             origin = '%s,%s' % (origin, self.name)
         return origin
 
@@ -1331,6 +1354,8 @@ class MrpProduction(models.Model):
                 # procurement and assigning is now run in write
                 move.write({'product_uom_qty': new_qty})
                 update_info.append((move, old_qty, new_qty))
+            if move.reference_ids != self.reference_ids:
+                move.reference_ids = self.reference_ids.ids
         return update_info
 
     @api.ondelete(at_uninstall=False)
@@ -1391,17 +1416,11 @@ class MrpProduction(models.Model):
 
     def _get_children(self):
         self.ensure_one()
-        procurement_moves = self.procurement_group_id.stock_move_ids
-        child_moves = procurement_moves.move_orig_ids
-        return ((procurement_moves | child_moves).created_production_id.procurement_group_id.mrp_production_ids\
-                | child_moves.production_id)\
-                .filtered(lambda p: p.origin != self.origin) - self
+        return self.production_group_id.child_ids.production_ids
 
     def _get_sources(self):
         self.ensure_one()
-        dest_moves = self.procurement_group_id.mrp_production_ids.move_dest_ids
-        parent_moves = self.procurement_group_id.stock_move_ids.move_dest_ids
-        return (dest_moves | parent_moves).group_id.mrp_production_ids.filtered(lambda p: p.origin != self.origin) - self
+        return self.production_group_id.parent_ids.production_ids
 
     def set_qty_producing(self):
         self.ensure_one()
@@ -1448,7 +1467,7 @@ class MrpProduction(models.Model):
         return action
 
     def action_view_mrp_production_backorders(self):
-        backorder_ids = self.procurement_group_id.mrp_production_ids.ids
+        backorder_ids = self.production_group_id.production_ids.ids
         return {
             'res_model': 'mrp.production',
             'type': 'ir.actions.act_window',
@@ -1531,12 +1550,12 @@ class MrpProduction(models.Model):
         moves_to_confirm = self.env['stock.move'].browse(sorted(moves_ids_to_confirm))
         workorder_to_confirm = self.env['mrp.workorder'].browse(sorted(workorder_ids_to_confirm))
 
+        ignored_mo_ids = self.env.context.get('ignore_mo_ids', [])
         move_raws_to_adjust._adjust_procure_method()
         moves_to_confirm._action_confirm(merge=False, create_proc=not self.env.context.get('no_procurement'))
         workorder_to_confirm._action_confirm()
         workorder_to_confirm._set_cost_mode()
         # run scheduler for moves forecasted to not have enough in stock
-        ignored_mo_ids = self.env.context.get('ignore_mo_ids', [])
         self.move_raw_ids.with_context(ignore_mo_ids=ignored_mo_ids + self.ids)._trigger_scheduler()
         self.picking_ids.filtered(
             lambda p: p.state not in ['cancel', 'done']).action_confirm()
@@ -1829,17 +1848,15 @@ class MrpProduction(models.Model):
             return name
         seq_back = "-" + "0" * (SIZE_BACK_ORDER_NUMERING - 1 - int(math.log10(sequence))) + str(sequence)
         regex = re.compile(r"-\d+$")
-        if regex.search(name) and (max(self.procurement_group_id.mrp_production_ids.mapped("backorder_sequence")) > 1 or sequence > 1):
+        if regex.search(name) and (max(self.production_group_id.production_ids.mapped("backorder_sequence")) > 1 or sequence > 1):
             return regex.sub(seq_back, name)
         return name + seq_back
 
     def _get_backorder_mo_vals(self):
         self.ensure_one()
-        if not self.procurement_group_id:
-            # in the rare case that the procurement group has been removed somehow, create a new one
-            self.procurement_group_id = self.env["procurement.group"].create({'name': self.name})
         return {
-            'procurement_group_id': self.procurement_group_id.id,
+            'reference_ids': self.reference_ids.ids,
+            'production_group_id': self.production_group_id.id,
             'move_raw_ids': None,
             'move_finished_ids': None,
             'lot_producing_ids': False,
@@ -1897,7 +1914,7 @@ class MrpProduction(models.Model):
             backorder_qtys = amounts[production][1:]
             production.with_context(skip_compute_move_raw_ids=True).product_qty = amounts[production][0]
 
-            next_seq = max(production.procurement_group_id.mrp_production_ids.mapped("backorder_sequence"), default=1)
+            next_seq = max(production.production_group_id.production_ids.mapped("backorder_sequence"), default=1)
 
             for qty_to_backorder in backorder_qtys:
                 next_seq += 1
@@ -2423,9 +2440,19 @@ class MrpProduction(models.Model):
             'product_qty': sum(production.product_uom_qty for production in self),
             'product_uom_id': product_id.uom_id.id,
             'user_id': user_id.id,
+            'reference_ids': [Command.link(r.id) for r in self.reference_ids],
             'origin': ",".join(sorted([production.name for production in self])),
         })
 
+        # update linked picking
+        self.env['stock.move'].search([
+            ('production_group_id', 'in', self.production_group_id.ids),
+        ]).production_group_id = production.production_group_id
+
+        # update linked mos
+        production.production_group_id.parent_ids = [Command.set(self.production_group_id.parent_ids.ids)]
+        production.production_group_id.child_ids = [Command.set(self.production_group_id.child_ids.ids)]
+
         for move in production.move_raw_ids:
             for field, vals in origs[move.bom_line_id.id].items():
                 move[field] = vals
@@ -2435,14 +2462,13 @@ class MrpProduction(models.Model):
 
         self.move_dest_ids.created_production_id = production.id
 
-        self.procurement_group_id.stock_move_ids.group_id = production.procurement_group_id
-
         if 'confirmed' in self.mapped('state'):
             production.move_raw_ids._adjust_procure_method()
             (production.move_raw_ids | production.move_finished_ids).write({'state': 'confirmed'})
             production.action_confirm()
 
         self.with_context(skip_activity=True)._action_cancel()
+        self.sudo().production_group_id.unlink()
         # set the new deadline of origin moves (stock to pre prod)
         production.move_raw_ids.move_orig_ids.with_context(date_deadline_propagate_ids=set(production.move_raw_ids.ids)).write({'date_deadline': production.date_start})
         for p in self:
@@ -2629,10 +2655,6 @@ class MrpProduction(models.Model):
         workorders_to_unlink.unlink()
         self.bom_id = bom
 
-    @api.model
-    def _prepare_procurement_group_vals(self, values):
-        return {'name': values['name']}
-
     def _get_quantity_to_backorder(self):
         self.ensure_one()
         return max(self.product_qty - self.qty_producing, 0)
@@ -3014,7 +3036,9 @@ class MrpProduction(models.Model):
 
     def _post_run_manufacture(self, post_production_values):
         note_subtype_id = self.env['ir.model.data']._xmlid_to_res_id('mail.mt_note')
-        for production in self:
+        for production, procurement in zip(self, post_production_values):
+            if group_id := procurement.values.get('production_group_id'):
+                production.production_group_id.parent_ids = [Command.link(group_id)]
             orderpoint = production.orderpoint_id
             origin_production = production.move_dest_ids.raw_material_production_id
             if orderpoint and orderpoint.create_uid.id == api.SUPERUSER_ID and orderpoint.trigger == 'manual':

--- a/addons/mrp/models/stock_move.py
+++ b/addons/mrp/models/stock_move.py
@@ -9,11 +9,31 @@ from odoo.exceptions import ValidationError
 class StockMove(models.Model):
     _inherit = 'stock.move'
 
+    @api.model
+    def default_get(self, fields):
+        defaults = super().default_get(fields)
+        if self.env.context.get('default_raw_material_production_id') or self.env.context.get('default_production_id'):
+            production_id = self.env['mrp.production'].browse(self.env.context.get('default_raw_material_production_id') or self.env.context.get('default_production_id'))
+
+            if production_id.state not in ('draft', 'cancel'):
+                if production_id.state != 'done':
+                    defaults['state'] = 'draft'
+                else:
+                    defaults['state'] = 'done'
+                    defaults['additional'] = True
+                defaults['product_uom_qty'] = 0.0
+            elif production_id.state == 'draft':
+                defaults['reference_ids'] = production_id.reference_ids.ids
+                defaults['reference'] = production_id.name
+        return defaults
+
     created_production_id = fields.Many2one('mrp.production', 'Created Production Order', check_company=True, index=True)
     production_id = fields.Many2one(
         'mrp.production', 'Production Order for finished products', check_company=True, index='btree_not_null', ondelete="cascade")
     raw_material_production_id = fields.Many2one(
         'mrp.production', 'Production Order for components', check_company=True, index='btree_not_null', ondelete="cascade")
+    production_group_id = fields.Many2one(
+        'mrp.production.group', 'Used for Productions')
     unbuild_id = fields.Many2one(
         'mrp.unbuild', 'Disassembly Order', check_company=True, index='btree_not_null')
     consume_unbuild_id = fields.Many2one(
@@ -154,6 +174,15 @@ class StockMove(models.Model):
                 moves_with_reference |= move
         super(StockMove, self - moves_with_reference)._compute_reference()
 
+    def _set_references(self):
+        super()._set_references()
+        for move in self:
+            if move.reference_ids:
+                continue
+            production = move.raw_material_production_id or move.production_id
+            if production:
+                move.reference_ids = [Command.set(production.reference_ids.ids)]
+
     @api.depends('raw_material_production_id.qty_producing', 'product_uom_qty', 'product_uom')
     def _compute_should_consume_qty(self):
         for move in self:
@@ -199,23 +228,6 @@ class StockMove(models.Model):
             if move.raw_material_production_id and move.product_uom.compare(move.quantity, 0) < 0:
                 raise ValidationError(_("Please enter a positive quantity."))
 
-    @api.model
-    def default_get(self, fields):
-        defaults = super().default_get(fields)
-        if self.env.context.get('default_raw_material_production_id') or self.env.context.get('default_production_id'):
-            production_id = self.env['mrp.production'].browse(self.env.context.get('default_raw_material_production_id') or self.env.context.get('default_production_id'))
-            if production_id.state not in ('draft', 'cancel'):
-                if production_id.state != 'done':
-                    defaults['state'] = 'draft'
-                else:
-                    defaults['state'] = 'done'
-                    defaults['additional'] = True
-                defaults['product_uom_qty'] = 0.0
-            elif production_id.state == 'draft':
-                defaults['group_id'] = production_id.procurement_group_id.id
-                defaults['reference'] = production_id.name
-        return defaults
-
     @api.model_create_multi
     def create(self, vals_list):
         """ Enforce consistent values (i.e. match _get_move_raw_values/_get_move_finished_values) for:
@@ -238,8 +250,9 @@ class StockMove(models.Model):
                     mo = mo.browse(mo_id)
                     mo_id_to_mo[mo_id] = mo
                 values['origin'] = mo._get_origin()
-                values['group_id'] = mo.procurement_group_id.id
                 values['propagate_cancel'] = mo.propagate_cancel
+                values['reference_ids'] = mo.reference_ids.ids
+                values['production_group_id'] = mo.production_group_id.id
                 if values.get('raw_material_production_id', False):
                     product = product_id_to_product[values['product_id']]
                     if not product:
@@ -311,14 +324,13 @@ class StockMove(models.Model):
                 possible_reduceable_qty = -sum(move.move_orig_ids.filtered(lambda m: m.state not in ('done', 'cancel') and m.product_uom_qty).mapped('product_uom_qty'))
                 procurement_qty = max(procurement_qty, possible_reduceable_qty)
                 values = move._prepare_procurement_values()
-                origin = move._prepare_procurement_origin()
-                procurements.append(self.env['procurement.group'].Procurement(
+                procurements.append(self.env['stock.rule'].Procurement(
                     move.product_id, procurement_qty, move.product_uom,
-                    move.location_id, move.reference, origin, move.company_id, values))
+                    move.location_id, move.reference, move.origin, move.company_id, values))
 
         to_assign._action_assign()
         if procurements:
-            self.env['procurement.group'].run(procurements)
+            self.env['stock.rule'].run(procurements)
 
     def _action_assign(self, force_qty=False):
         res = super(StockMove, self)._action_assign(force_qty=force_qty)
@@ -502,7 +514,7 @@ class StockMove(models.Model):
     @api.model
     def _prepare_merge_moves_distinct_fields(self):
         res = super()._prepare_merge_moves_distinct_fields()
-        res += ['created_production_id', 'cost_share']
+        res += ['created_production_id', 'cost_share', 'production_group_id']
         if self.bom_line_id and ("phantom" in self.bom_line_id.bom_id.mapped('type')):
             res.append('bom_line_id')
         return res
@@ -575,9 +587,15 @@ class StockMove(models.Model):
 
     def _prepare_procurement_values(self):
         res = super()._prepare_procurement_values()
+        res['production_group_id'] = self.production_group_id.id
         res['bom_line_id'] = self.bom_line_id.id
         return res
 
+    def _search_picking_for_assignation_domain(self):
+        domain = super()._search_picking_for_assignation_domain()
+        domain += [('production_group_id', '=', self.production_group_id.id)]
+        return domain
+
     def action_open_reference(self):
         res = super().action_open_reference()
         source = self.production_id or self.raw_material_production_id

--- a/addons/mrp/models/stock_orderpoint.py
+++ b/addons/mrp/models/stock_orderpoint.py
@@ -213,8 +213,8 @@ class StockWarehouseOrderpoint(models.Model):
                         prod.product_qty, orderpoint.product_uom, round=False)
         return res
 
-    def _prepare_procurement_values(self, date=False, group=False):
-        values = super()._prepare_procurement_values(date=date, group=group)
+    def _prepare_procurement_values(self, date=False):
+        values = super()._prepare_procurement_values(date=date)
         values['bom_id'] = self.bom_id
         return values
 

--- a/addons/mrp/models/stock_picking.py
+++ b/addons/mrp/models/stock_picking.py
@@ -132,23 +132,27 @@ class StockPicking(models.Model):
         compute='_compute_mrp_production_ids',
         groups='mrp.group_mrp_user')
 
-    production_ids = fields.Many2many(
+    production_ids = fields.One2many(
         'mrp.production',
-        compute='_compute_mrp_production_ids',
+        related='move_ids.production_group_id.production_ids',
         groups='mrp.group_mrp_user')
+    production_group_id = fields.Many2one(
+        'mrp.production.group',
+        string="Production Group",
+        related='move_ids.production_group_id',
+    )
 
     @api.depends('move_ids')
     def _compute_has_kits(self):
         for picking in self:
             picking.has_kits = any(picking.move_ids.mapped('bom_line_id'))
 
-    @api.depends('group_id')
+    @api.depends('production_ids')
     def _compute_mrp_production_ids(self):
         for picking in self:
-            production_ids = picking.group_id.mrp_production_ids | picking.move_ids.move_dest_ids.raw_material_production_id
-            # Filter out unwanted MO types
-            picking.production_ids = production_ids.filtered(lambda p: p.picking_type_id.active)
-            picking.production_count = len(picking.production_ids)
+            # hide subcontracting MO from resupply picking
+            mo = picking.production_ids.filtered(lambda mo: mo.picking_type_id.active)
+            picking.production_count = len(mo)
 
     def action_detailed_operations(self):
         action = super().action_detailed_operations()

--- a/None
+++ b/addons/mrp/models/stock_reference.py
@@ -0,0 +1,9 @@
+from odoo import fields, models
+
+
+class StockReference(models.Model):
+    _inherit = 'stock.reference'
+
+    production_ids = fields.Many2many(
+        'mrp.production', 'stock_reference_production_rel', 'reference_id',
+        'production_id', string="Productions")

--- a/addons/mrp/models/stock_rule.py
+++ b/addons/mrp/models/stock_rule.py
@@ -5,7 +5,7 @@ from datetime import datetime
 from dateutil.relativedelta import relativedelta
 
 from odoo import api, fields, models, SUPERUSER_ID, _
-from odoo.fields import Domain
+from odoo.fields import Domain, Command
 from odoo.tools import OrderedSet
 
 
@@ -37,6 +37,39 @@ class StockRule(models.Model):
             return (not p.workorder_ids and (p.orderpoint_id or p.move_dest_ids.procure_method == 'make_to_stock'))
         return not p.orderpoint_id
 
+    @api.model
+    def run(self, procurements, raise_user_error=True):
+        """ If 'run' is called on a kit, this override is made in order to call
+        the original 'run' method with the values of the components of that kit.
+        """
+        procurements_without_kit = []
+        product_by_company = defaultdict(OrderedSet)
+        for procurement in procurements:
+            product_by_company[procurement.company_id].add(procurement.product_id.id)
+        kits_by_company = {
+            company: self.env['mrp.bom']._bom_find(self.env['product.product'].browse(product_ids), company_id=company.id, bom_type='phantom')
+            for company, product_ids in product_by_company.items()
+        }
+        for procurement in procurements:
+            bom_kit = kits_by_company[procurement.company_id].get(procurement.product_id)
+            if bom_kit:
+                order_qty = procurement.product_uom._compute_quantity(procurement.product_qty, bom_kit.product_uom_id, round=False)
+                qty_to_produce = (order_qty / bom_kit.product_qty)
+                _dummy, bom_sub_lines = bom_kit.explode(procurement.product_id, qty_to_produce, never_attribute_values=procurement.values.get("never_product_template_attribute_value_ids"))
+                for bom_line, bom_line_data in bom_sub_lines:
+                    bom_line_uom = bom_line.product_uom_id
+                    quant_uom = bom_line.product_id.uom_id
+                    # recreate dict of values since each child has its own bom_line_id
+                    values = dict(procurement.values, bom_line_id=bom_line.id)
+                    component_qty, procurement_uom = bom_line_uom._adjust_uom_quantities(bom_line_data['qty'], quant_uom)
+                    procurements_without_kit.append(self.env['stock.rule'].Procurement(
+                        bom_line.product_id, component_qty, procurement_uom,
+                        procurement.location_id, procurement.name,
+                        procurement.origin, procurement.company_id, values))
+            else:
+                procurements_without_kit.append(procurement)
+        return super().run(procurements_without_kit, raise_user_error=raise_user_error)
+
     @api.model
     def _run_manufacture(self, procurements):
         new_productions_values_by_company = defaultdict(lambda: defaultdict(list))
@@ -72,42 +105,20 @@ class StockRule(models.Model):
             productions_vals_list = new_productions_values_by_company[company_id]['values']
             # create the MO as SUPERUSER because the current user may not have the rights to do it (mto product launched by a sale for example)
             productions = self.env['mrp.production'].with_user(SUPERUSER_ID).sudo().with_company(company_id).create(productions_vals_list)
-            productions.filtered(self._should_auto_confirm_procurement_mo).action_confirm()
+            for mo in productions:
+                if self._should_auto_confirm_procurement_mo(mo):
+                    mo.action_confirm()
             productions._post_run_manufacture(new_productions_values_by_company[company_id]['procurements'])
         return True
 
-    @api.model
-    def _run_pull(self, procurements):
-        # Override to correctly assign the move generated from the pull
-        # in its production order (pbm_sam only)
-        for procurement, rule in procurements:
-            warehouse_id = rule.warehouse_id
-            if not warehouse_id:
-                warehouse_id = rule.location_dest_id.warehouse_id
-            manu_rule = rule.route_id.rule_ids.filtered(lambda r: r.action == 'manufacture' and r.warehouse_id == warehouse_id)
-            if warehouse_id.manufacture_steps != 'pbm_sam' or not manu_rule:
-                continue
-            if rule.picking_type_id == warehouse_id.sam_type_id or (
-                warehouse_id.sam_loc_id and warehouse_id.sam_loc_id.parent_path in rule.location_src_id.parent_path
-            ):
-                if procurement.product_uom.compare(procurement.product_qty, 0) < 0:
-                    procurement.values['group_id'] = procurement.values['group_id'].stock_move_ids.filtered(
-                        lambda m: m.state not in ['done', 'cancel']).move_orig_ids.group_id[:1]
-                    continue
-                manu_type_id = manu_rule[0].picking_type_id
-                if manu_type_id:
-                    name = manu_type_id.sequence_id.next_by_id()
-                else:
-                    name = self.env['ir.sequence'].next_by_code('mrp.production') or _('New')
-                # Create now the procurement group that will be assigned to the new MO
-                # This ensure that the outgoing move PostProduction -> Stock is linked to its MO
-                # rather than the original record (MO or SO)
-                group = procurement.values.get('group_id')
-                if group:
-                    procurement.values['group_id'] = group.copy({'name': name})
-                else:
-                    procurement.values['group_id'] = self.env["procurement.group"].create({'name': name})
-        return super()._run_pull(procurements)
+    def _get_stock_move_values(self, product_id, product_qty, product_uom, location_id, name, origin, company_id, values):
+        res = super()._get_stock_move_values(product_id, product_qty, product_uom, location_id, name, origin, company_id, values)
+        res['production_group_id'] = values.get('production_group_id')
+        return res
+
+    def _get_moves_to_assign_domain(self, company_id):
+        domain = super()._get_moves_to_assign_domain(company_id)
+        return Domain(domain) & Domain('production_id', '=', False)
 
     def _get_custom_move_fields(self):
         fields = super(StockRule, self)._get_custom_move_fields()
@@ -125,9 +136,6 @@ class StockRule(models.Model):
         return self.env['mrp.bom']._bom_find(product_id, picking_type=False, bom_type='normal', company_id=company_id.id)[product_id]
 
     def _make_mo_get_domain(self, procurement, bom):
-        gpo = self.group_propagation_option
-        group = (gpo == 'fixed' and self.group_id) or \
-                (gpo == 'propagate' and 'group_id' in procurement.values and procurement.values['group_id']) or False
         domain = (
             ('bom_id', '=', bom.id),
             ('product_id', '=', procurement.product_id.id),
@@ -136,6 +144,7 @@ class StockRule(models.Model):
             ('picking_type_id', '=', self.picking_type_id.id),
             ('company_id', '=', procurement.company_id.id),
             ('user_id', '=', False),
+            ('reference_ids', '=', procurement.values.get('reference_ids', self.env['stock.reference']).ids),
         )
         if procurement.values.get('orderpoint_id'):
             procurement_date = datetime.combine(
@@ -145,8 +154,6 @@ class StockRule(models.Model):
             domain += ('|',
                        '&', ('state', '=', 'draft'), ('date_deadline', '<=', procurement_date),
                        '&', ('state', '=', 'confirmed'), ('date_start', '<=', procurement_date))
-        if group:
-            domain += (('procurement_group_id', '=', group.id),)
         return domain
 
     def _prepare_mo_vals(self, product_id, product_qty, product_uom, location_dest_id, name, origin, company_id, values, bom):
@@ -166,7 +173,7 @@ class StockRule(models.Model):
             'bom_id': bom.id,
             'date_deadline': date_deadline,
             'date_start': date_planned,
-            'procurement_group_id': False,
+            'reference_ids': [Command.set(values.get('reference_ids', self.env['stock.reference']).ids)],
             'propagate_cancel': self.propagate_cancel,
             'orderpoint_id': values.get('orderpoint_id', False) and values.get('orderpoint_id').id,
             'picking_type_id': picking_type.id or values['warehouse_id'].manu_type_id.id,
@@ -174,15 +181,6 @@ class StockRule(models.Model):
             'move_dest_ids': values.get('move_dest_ids') and [(4, x.id) for x in values['move_dest_ids']] or False,
             'user_id': False,
         }
-        # Use the procurement group created in _run_pull mrp override
-        # Preserve the origin from the original stock move, if available
-        if location_dest_id.warehouse_id.manufacture_steps == 'pbm_sam' and values.get('move_dest_ids') and values.get('group_id') and values['group_id'].name not in values['move_dest_ids'][0].origin:
-            origin = values['move_dest_ids'][0].origin
-            mo_values.update({
-                'name': values['group_id'].name,
-                'procurement_group_id': values['group_id'].id,
-                'origin': origin,
-            })
         if self.location_dest_from_rule:
             mo_values['location_dest_id'] = self.location_dest_id.id
         return mo_values
@@ -229,50 +227,7 @@ class StockRule(models.Model):
         return delays, delay_description
 
     def _push_prepare_move_copy_values(self, move_to_copy, new_date):
-        new_move_vals = super(StockRule, self)._push_prepare_move_copy_values(move_to_copy, new_date)
+        new_move_vals = super()._push_prepare_move_copy_values(move_to_copy, new_date)
+        new_move_vals['production_group_id'] = move_to_copy.production_group_id.id
         new_move_vals['production_id'] = False
         return new_move_vals
-
-
-class ProcurementGroup(models.Model):
-    _inherit = 'procurement.group'
-
-    mrp_production_ids = fields.One2many('mrp.production', 'procurement_group_id')
-
-    @api.model
-    def run(self, procurements, raise_user_error=True):
-        """ If 'run' is called on a kit, this override is made in order to call
-        the original 'run' method with the values of the components of that kit.
-        """
-        procurements_without_kit = []
-        product_by_company = defaultdict(OrderedSet)
-        for procurement in procurements:
-            product_by_company[procurement.company_id].add(procurement.product_id.id)
-        kits_by_company = {
-            company: self.env['mrp.bom']._bom_find(self.env['product.product'].browse(product_ids), company_id=company.id, bom_type='phantom')
-            for company, product_ids in product_by_company.items()
-        }
-        for procurement in procurements:
-            bom_kit = kits_by_company[procurement.company_id].get(procurement.product_id)
-            if bom_kit:
-                order_qty = procurement.product_uom._compute_quantity(procurement.product_qty, bom_kit.product_uom_id, round=False)
-                qty_to_produce = (order_qty / bom_kit.product_qty)
-                _dummy, bom_sub_lines = bom_kit.explode(procurement.product_id, qty_to_produce, never_attribute_values=procurement.values.get("never_product_template_attribute_value_ids"))
-                for bom_line, bom_line_data in bom_sub_lines:
-                    bom_line_uom = bom_line.product_uom_id
-                    quant_uom = bom_line.product_id.uom_id
-                    # recreate dict of values since each child has its own bom_line_id
-                    values = dict(procurement.values, bom_line_id=bom_line.id)
-                    component_qty, procurement_uom = bom_line_uom._adjust_uom_quantities(bom_line_data['qty'], quant_uom)
-                    procurements_without_kit.append(self.env['procurement.group'].Procurement(
-                        bom_line.product_id, component_qty, procurement_uom,
-                        procurement.location_id, procurement.name,
-                        procurement.origin, procurement.company_id, values))
-            else:
-                procurements_without_kit.append(procurement)
-        return super().run(procurements_without_kit, raise_user_error=raise_user_error)
-
-    def _get_moves_to_assign_domain(self, company_id):
-        domain = super()._get_moves_to_assign_domain(company_id)
-        domain = Domain.AND([domain, Domain('production_id', '=', False)])
-        return domain

--- a/addons/mrp/models/stock_scrap.py
+++ b/addons/mrp/models/stock_scrap.py
@@ -85,8 +85,8 @@ class StockScrap(models.Model):
     def do_replenish(self, values=False):
         self.ensure_one()
         values = values or {}
-        if self.production_id and self.production_id.procurement_group_id:
+        if self.production_id and self.production_id.production_group_id:
             values.update({
-                'group_id': self.production_id.procurement_group_id,
+                'production_group_id': self.production_id.production_group_id.id,
             })
         super().do_replenish(values)
